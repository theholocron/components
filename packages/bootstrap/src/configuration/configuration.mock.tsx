import * as React from "react";
import { useConfiguration } from "./configuration";

export const mockConf = {
	application: {
		description: "This is a description for a mock application.",
		id: "mock-app-id",
		name: "A Mock Application",
		get slug (): string {
			return `${this.id}-slug`;
		},
	},
	// fetch: {},
};

export function TestComponent () {
	const conf = useConfiguration();

	return (
		<div data-testid="mock-configuration">
			<pre>{JSON.stringify(conf, null, 4)}</pre>
		</div>
	);
}
