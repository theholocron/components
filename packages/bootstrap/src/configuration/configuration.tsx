import * as str from "@theholocron/utils-string";
import * as React from "react";
import type { WithChildren } from "../type.ts";

export interface ApplicationConf {
	description?: string;
	id: string;
	name: string;
	slug?: string;
	title?: string;
	version?: string;
}

export interface IConfiguration {
	application: ApplicationConf;
	// fetch?: FetchConf;
}

const Context = React.createContext<IConfiguration | undefined>(undefined);
Context.displayName = "@theholocron/bootstrap/Configuration";

export interface ConfigurationProps extends WithChildren {
	conf: IConfiguration;
}

export function Provider(props: ConfigurationProps) {
	const conf = {
		...props.conf,
		application: {
			...props.conf.application,
			slug: str.toKebabCase(
				props.conf.application?.slug || props.conf.application.id
			),
			title:
				props.conf.application?.title ??
				str.toTitleCase(props.conf.application?.name),
		},
	};

	const value = React.useMemo(() => conf, [conf]);

	return <Context.Provider value={value}>{props.children}</Context.Provider>;
}
Provider.displayName = "@theholocron/bootstrap/Configuration";

export function useConfiguration(): IConfiguration {
	const context = React.useContext(Context);

	if (context === undefined) {
		throw new Error(
			`useConfiguration must be used within Configuration.Provider!`
		);
	}

	return context;
}
