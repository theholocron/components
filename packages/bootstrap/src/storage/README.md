# Browser Storage

A React Provider component for storing data within browser storage utilizing `window.sessionStorage`.

## Usage

```typescript
import { Storage } from "@theholocron/bootstrap";
import * as React from "react";

const conf: ApplicationConf = {
    application: { id: "mock-app-id" },
};

interface AppProps {
    Component: React.ReactNode;
    pageProps?: any;
}

export default function App (props: AppProps) {
    const { Component, pageProps } = props;

    return (
        <Storage.Provider conf={conf.application}>
            <Component {...pageProps} />
        </Storage.Provider>;
    );
}
```

### `useStorage`

Use this function for getting, setting and removing items from the storage, i.e. `sessionStorage`.

```typescript
import { useStorage } from "@theholocron/bootstrap";

const storage = useStorage();

// send an item to the store
storage.sendTo("mockKey", item);

// grab an item from the store
const item = storage.getFrom("mockKey");

// delete it if you want
storage.removeFrom("mockKey");
```
