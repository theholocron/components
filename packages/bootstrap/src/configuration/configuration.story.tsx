import { expect, userEvent, within } from "@storybook/test";
import * as React from "react";
import * as MockConfiguration from "./configuration.mock";
import Configuration, { useConfiguration } from "./index";

export default {
	argTypes: {
		conf: {
			control: "object",
			defaultValue: MockConfiguration.mockConf,
			description: "The applications configuration."
		},
	},
	component: Configuration.Provider,
	title: "Configuration",
};

export const Default = (args) => (
	<Configuration.Provider {...args}>
		<MockConfiguration.TestComponent />
	</Configuration.Provider>
);
Default.args = {
	conf: MockConfiguration.mockConf,
}
Default.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);

	// Check if the initial storage state is as expected
	const initialStorage = await canvas.findByTestId("mock-configuration");
	expect(initialStorage).toBeTruthy();

	const appId = await canvas.findByText(/"id": "mock-app-id"/);
	expect(appId).toBeInTheDocument();

	const appName = await canvas.findByText(/"name": "A Mock Application"/);
	expect(appName).toBeInTheDocument();
};

// Attempt to use the hook without provider
export const Error = () => {
	const Component = () => {
		try {
			useConfiguration();
		} catch (error) {
			return <div data-testid="error-message">{error.message}</div>;
		}
		return null; // Render nothing if no error
	};

	return <Component />;
};
Error.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);

	const errorMessage = await canvas.findByTestId("error-message");
	expect(errorMessage).toHaveTextContent("useConfiguration must be used within Configuration.Provider!");
};

// Fallback to ID when slug is not present
export const FallbackSlug = Default.bind({});
FallbackSlug.args = {
	conf: {
		...MockConfiguration.mockConf,
		application: {
			...MockConfiguration.mockConf.application,
			name: "",
			id: "fallback-slug",
			slug: "",
		},
	},
};
FallbackSlug.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);
	const testComponent = await canvas.findByTestId("mock-configuration");
	const appName = await canvas.findByText(/"slug": "fallback-slug"/);
	expect(appName).toBeInTheDocument();
};

// Fallback when only ID is present
export const FallbackTitle = Default.bind({});
FallbackTitle.args = {
	conf: MockConfiguration.mockConf,
};
FallbackTitle.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);
	const testComponent = await canvas.findByTestId("mock-configuration");
	const appName = await canvas.findByText(/"title": "A Mock Application"/);
	expect(appName).toBeInTheDocument();
};
