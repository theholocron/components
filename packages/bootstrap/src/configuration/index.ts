import { Provider } from "./configuration";

export { useConfiguration } from "./configuration";
export type { ApplicationConf, IConfiguration } from "./configuration";

export default { Provider };
