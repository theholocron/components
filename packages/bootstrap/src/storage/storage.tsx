import { storage, type TSessionStorage } from "@theholocron/utils-storage";
import * as React from "react";
import { type ApplicationConf } from "../configuration/index";
import type { WithChildren } from "../type";

const Context = React.createContext<TSessionStorage | undefined>(undefined);
Context.displayName = "@theholocron/bootstrap/Storage";

export interface StorageProps extends WithChildren {
	conf: ApplicationConf;
}

export function Provider(props: StorageProps) {
	const value = React.useMemo(() => {
		const vault = storage.session.create();

		vault.registerApp(props.conf.id);
		vault.sendTo("application.name", props.conf.name);
		vault.sendTo("build.version", props.conf?.version ?? "0.0.0");

		return vault;
	}, [props.conf.id, props.conf.name, props.conf.version]);

	return <Context.Provider value={value}>{props.children}</Context.Provider>;
}
Provider.displayName = "@theholocron/bootstrap/Storage";

export function useStorage(): TSessionStorage {
	const context = React.useContext(Context);

	if (context === undefined) {
		throw new Error("useStorage must be used within Storage.Provider!");
	}

	return context;
}
