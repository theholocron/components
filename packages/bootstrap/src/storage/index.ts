import { type TSessionStorage } from "@theholocron/utils-storage";
import { Provider } from "./storage";

export { useStorage } from "./storage";
export type { TSessionStorage } from "./storage";

export default { Provider };
