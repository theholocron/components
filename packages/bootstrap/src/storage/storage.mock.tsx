import * as React from "react";
import { useStorage } from "./storage";

export function TestComponent () {
	const [isSending, setIsSending] = React.useState(true);
	const [data, setData] = React.useState({});
	const storage = useStorage();

	const handleClick = () => {
		if (isSending) {
			storage.sendTo("test", "value");
		} else {
			storage.removeFrom("test");
		}
		setData(storage.getAll());
		setIsSending(!isSending);
	};

	React.useEffect(() => {
		setData(JSON.parse(window.sessionStorage["@theholocron"]));
	}, []);

	return (
		<div data-testid="mock-storage">
			<button
				onClick={handleClick}
				type="button"
			>
				{isSending ? "Send to Storage" : "Remove from Storage"}
			</button>
			<pre>{JSON.stringify(data, null, 4)}</pre>
		</div>
	);
}
