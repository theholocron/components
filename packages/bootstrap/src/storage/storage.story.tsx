import { expect, userEvent, within } from "@storybook/test";
import * as React from "react";
import * as MockConfiguration from "../configuration/configuration.mock";
import Storage, { useStorage } from "./index";
import * as MockStorage from "./storage.mock";

export default {
	argTypes: {
		conf: {
			control: "object",
			description: "The applications configuration object, specifically the application part of it."
		},
	},
	component: Storage.Provider,
	title: "Storage",
};

export const Default = (args) => (
	<Storage.Provider {...args}>
		<MockStorage.TestComponent />
	</Storage.Provider>
);
Default.args = {
	conf: MockConfiguration.mockConf.application,
};
Default.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);

	// Check if the initial storage state is as expected
	const initialStorage = await canvas.findByTestId("mock-storage");
	expect(initialStorage).toBeTruthy();

	// 1. Find the button to send to storage
	const sendButton = await canvas.findByRole('button', { name: /send to storage/i });
	// 2. Click the button to send value to storage
	await userEvent.click(sendButton);
	// 3. After clicking, verify that the storage state has been updated
	const updatedDataDisplay = await canvas.findByText(/"test": "value"/);
	expect(updatedDataDisplay).toBeInTheDocument();

	// 1. Find the button to remove from storage
	const removeButton = await canvas.findByRole('button', { name: /remove from storage/i });
	// 2. Click the button to remove value from storage
	await userEvent.click(removeButton);
	// 3. After clicking, verify that the storage state has been updated
	const preElement = await canvas.queryByText(/"test": "value"/);
	expect(preElement).not.toBeInTheDocument();
};

// Attempt to use the hook without provider
export const Error = () => {
	const Component = () => {
		try {
			useStorage();
		} catch (error) {
			return <div data-testid="error-message">{error.message}</div>;
		}
		return null; // Render nothing if no error
	};

	return <Component />;
};
Error.args = {};
Error.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);

	const errorMessage = await canvas.findByTestId("error-message");
	expect(errorMessage).toHaveTextContent("useStorage must be used within Storage.Provider!");
};
