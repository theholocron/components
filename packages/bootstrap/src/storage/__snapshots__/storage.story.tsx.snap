// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storage Default play-test 1`] = `
<div data-testid="mock-storage">
  <button type="button">
    Send to Storage
  </button>
  <pre>
    {
    "@theholocron": {
        "mock-app-id": {
            "application": {
                "name": "A Mock Application"
            },
            "build": {
                "version": "0.0.0"
            }
        }
    }
}
  </pre>
</div>
`;

exports[`Storage Error play-test 1`] = `
<div data-testid="error-message">
  useStorage must be used within Storage.Provider!
</div>
`;
